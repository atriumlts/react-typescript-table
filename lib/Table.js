"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var change_case_1 = require("change-case");
function defaultTitle(key) {
    return change_case_1.titleCase("" + key);
}
function getColumnTitle(column) {
    var title = column.title === undefined ? defaultTitle(column.key) : column.title;
    return React.createElement("th", { key: "" + column.key }, title);
}
exports.getColumnTitle = getColumnTitle;
function defaultFormat(value) {
    if (value === null || value === undefined) {
        return '';
    }
    return "" + value;
}
var Table = (function (_super) {
    __extends(Table, _super);
    function Table() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isColumnDescriptor = function (value) {
            return _this.props.data[0].hasOwnProperty(value.key);
        };
        _this.getFormattedColumn = function (column, row) {
            var title = _this.isColumnDescriptor(column)
                ? !!column.format
                    ? column.format(row)
                    : defaultFormat(row[column.key])
                : column.format(row);
            var tdClassName = column.tdClassName || '';
            return (React.createElement("td", { className: tdClassName, key: "" + column.key }, title));
        };
        _this.rowColumns = function (row) {
            if (_this.props.columns == undefined && _this.props.data.length === 0) {
                return [];
            }
            if (_this.props.columns == undefined) {
                return Object.entries(row).map(function (_a) {
                    var key = _a[0], value = _a[1];
                    return (React.createElement("td", { key: "" + key }, defaultFormat(value)));
                });
            }
            else {
                return _this.props.columns.map(function (column) {
                    return _this.getFormattedColumn(column, row);
                });
            }
        };
        _this.formatRow = function (row, trAttributes, trClassName) {
            return (React.createElement("tr", __assign({ "data-key": row.id, key: "" + row.id }, trAttributes, { className: trClassName ? trClassName(row) : '' }), _this.rowColumns(row)));
        };
        return _this;
    }
    Object.defineProperty(Table.prototype, "columnTitles", {
        get: function () {
            if (this.props.columns == undefined && this.props.data.length === 0) {
                return [];
            }
            if (this.props.columns == undefined) {
                return Object.keys(this.props.data[0]).map(function (key) { return React.createElement("th", { key: "" + key }, defaultTitle(key)); });
            }
            else {
                return this.props.columns.map(getColumnTitle);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Table.prototype, "rows", {
        get: function () {
            var _this = this;
            return this.props.data.map(function (value) {
                return _this.formatRow(value, _this.props.trAttributes, _this.props.trClassName);
            });
        },
        enumerable: true,
        configurable: true
    });
    Table.prototype.render = function () {
        return (React.createElement("table", { className: this.props.tableClassName },
            React.createElement("thead", null,
                React.createElement("tr", null, this.columnTitles)),
            React.createElement("tbody", null, this.rows)));
    };
    return Table;
}(React.Component));
exports.default = Table;
//# sourceMappingURL=Table.js.map