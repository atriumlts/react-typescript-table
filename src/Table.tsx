import * as React from 'react'
import { titleCase } from 'change-case'

export type UniqueDataType = {
  id: string | number
}

export type Props<RowDataType, OtherColumns> = {
  data: Array<RowDataType>
  columns?: Array<ColumnDescription<RowDataType, OtherColumns>>
  minimumRowCount?: number
  tableClassName?: string
  trAttributes?: React.HTMLProps<HTMLTableRowElement>
  trClassName?: (row: RowDataType) => string
}

export type ColumnDescription<RowDataType, OtherColumns> =
  | ColumnDescriptor<RowDataType, keyof RowDataType>
  | OtherColumnDescriptor<RowDataType, keyof OtherColumns>

export type ColumnDescriptor<RowDataType, AttributeKey extends keyof RowDataType> = {
  key: AttributeKey
  title?: string | React.ReactNode
  tdClassName?: string
  format?: (value: RowDataType) => React.ReactText | React.ReactNode
}

export type OtherColumnDescriptor<RowDataType, OtherColumnKeys> = {
  key: OtherColumnKeys
  title?: string | React.ReactNode
  tdClassName?: string
  format: (value: RowDataType) => React.ReactText | React.ReactNode
}

function defaultTitle(key: any): string {
  return titleCase(`${key}`)
}

export function getColumnTitle<RowDataType, OtherColumns>(
  column: ColumnDescription<RowDataType, OtherColumns>
): React.ReactElement<'th'> {
  const title = column.title === undefined ? defaultTitle(column.key) : column.title

  return <th key={`${column.key}`}>{title}</th>
}

function defaultFormat(value: any): string {
  if (value === null || value === undefined) {
    return ''
  }
  return `${value}`
}

export default class Table<
  RowDataType extends UniqueDataType,
  OtherColumns
> extends React.Component<Props<RowDataType, OtherColumns>> {
  isColumnDescriptor = (
    value: ColumnDescription<RowDataType, OtherColumns>
  ): value is ColumnDescriptor<RowDataType, keyof RowDataType> => {
    return this.props.data[0].hasOwnProperty(value.key)
  }

  get columnTitles(): Array<React.ReactElement<'th'>> {
    if (this.props.columns == undefined && this.props.data.length === 0) {
      return []
    }
    if (this.props.columns == undefined) {
      return Object.keys(this.props.data[0]).map(key => <th key={`${key}`}>{defaultTitle(key)}</th>)
    } else {
      return this.props.columns.map(getColumnTitle)
    }
  }

  getFormattedColumn = (
    column: ColumnDescription<RowDataType, OtherColumns>,
    row: RowDataType
  ): React.ReactElement<'td'> => {
    const title = this.isColumnDescriptor(column)
      ? !!column.format
        ? column.format(row)
        : defaultFormat(row[column.key])
      : column.format(row)
    const tdClassName = column.tdClassName || ''
    return (
      <td className={tdClassName} key={`${column.key}`}>
        {title}
      </td>
    )
  }

  rowColumns = (row: RowDataType): Array<React.ReactElement<'td'>> => {
    if (this.props.columns == undefined && this.props.data.length === 0) {
      return []
    }
    if (this.props.columns == undefined) {
      return Object.entries(row).map(([key, value]: [string, any]) => (
        <td key={`${key}`}>{defaultFormat(value)}</td>
      ))
    } else {
      return this.props.columns.map((column: ColumnDescription<RowDataType, OtherColumns>) =>
        this.getFormattedColumn(column, row)
      )
    }
  }

  formatRow = (
    row: RowDataType,
    trAttributes?: React.HTMLProps<HTMLTableRowElement>,
    trClassName?: (row: RowDataType) => string
  ): React.ReactElement<'tr'> => {
    return (
      <tr
        data-key={row.id}
        key={`${row.id}`}
        {...trAttributes}
        className={trClassName ? trClassName(row) : ''}
      >
        {this.rowColumns(row)}
      </tr>
    )
  }

  get rows(): Array<React.ReactElement<'tr'>> {
    return this.props.data.map((value: RowDataType) => {
        return this.formatRow(value, this.props.trAttributes, this.props.trClassName)
      })
  }

  render() {
    return (
      <table className={this.props.tableClassName}>
        <thead>
          <tr>{this.columnTitles}</tr>
        </thead>
        <tbody>{this.rows}</tbody>
      </table>
    )
  }
}
