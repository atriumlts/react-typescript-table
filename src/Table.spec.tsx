import * as React from 'react'
import { mount, shallow } from 'enzyme'
import Table, { Props, ColumnDescription } from './Table'

const EXAMPLE_DATA = [
  {
    id: 1,
    name: 'Verla Bashirian',
    dayOfArrival: '1/2/18'
  },
  {
    id: 2,
    name: 'Daniella Mante',
    dayOfArrival: '2/16/17'
  },
  {
    id: 3,
    name: 'Estrella Feil',
    dayOfArrival: '5/12/18'
  },
  {
    id: 4,
    name: 'Hilbert McKenzie',
    dayOfArrival: '12/22/19'
  },
  {
    id: 5,
    name: 'Myah Kerluke',
    dayOfArrival: '9/7/15'
  }
]

type ExampleRowType = {
  id: number
  name: string
  dayOfArrival: string
}

const COLUMN_DESCRIPTIONS: Array<ColumnDescription<ExampleRowType, {}>> = [
  {
    key: 'dayOfArrival',
    format: (row: ExampleRowType): string => row.dayOfArrival.replace('/', ' ')
  },
  {
    key: 'name',
    title: 'Applicant Name'
  }
]

describe('Table', () => {
  describe('Without column descriptions', () => {
    function sampleComponent() {
      const props: Props<ExampleRowType, {}> = {
        data: EXAMPLE_DATA,
        columns: undefined
      }

      return <Table {...props} />
    }

    it('should match the snapshot', () => {
      const component = shallow(sampleComponent())
      expect(component).toMatchSnapshot()
    })
  })

  describe('With column descriptions', () => {
    function sampleComponent() {
      const props: Props<ExampleRowType, {}> = {
        data: EXAMPLE_DATA,
        columns: COLUMN_DESCRIPTIONS
      }

      return <Table {...props} />
    }

    it('should match the snapshot', () => {
      const component = shallow(sampleComponent())
      expect(component).toMatchSnapshot()
    })
  })

  describe('With an extra column', () => {
    function sampleComponent(
      mockedButtonCallback: Function = jest.fn()
    ) {
      const COLUMNS: Array<ColumnDescription<ExampleRowType, { action_delete: null }>> = [
        ...COLUMN_DESCRIPTIONS,
        {
          key: 'action_delete',
          title: 'Delete',
          format: (row: ExampleRowType): React.ReactNode => {
            return <button onClick={() => mockedButtonCallback(row.id)}>Delete</button>
          }
        }
      ]

      const props: Props<ExampleRowType, { action_delete: null }> = {
        data: EXAMPLE_DATA,
        columns: COLUMNS
      }

      return <Table {...props} />
    }

    it('should match the snapshot', () => {
      const component = shallow(sampleComponent())
      expect(component).toMatchSnapshot()
    })

    it('render a usable CTA', () => {
      const mockedButtonCallback = jest.fn()
      const component = mount(sampleComponent(mockedButtonCallback))

      const button = component
        .find('button')
        .find({ children: 'Delete' })
        .first()
      button.simulate('click')

      expect(mockedButtonCallback.mock.calls).toHaveLength(1)
    })
  })
})
