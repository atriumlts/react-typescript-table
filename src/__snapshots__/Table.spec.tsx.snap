// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table With an extra column should match the snapshot 1`] = `
<table>
  <thead>
    <tr>
      <th
        key="dayOfArrival"
      >
        Day Of Arrival
      </th>
      <th
        key="name"
      >
        Applicant Name
      </th>
      <th
        key="action_delete"
      >
        Delete
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className=""
      data-key={1}
      key="1"
    >
      <td
        className=""
        key="dayOfArrival"
      >
        1 2/18
      </td>
      <td
        className=""
        key="name"
      >
        Verla Bashirian
      </td>
      <td
        className=""
        key="action_delete"
      >
        <button
          onClick={[Function]}
        >
          Delete
        </button>
      </td>
    </tr>
    <tr
      className=""
      data-key={2}
      key="2"
    >
      <td
        className=""
        key="dayOfArrival"
      >
        2 16/17
      </td>
      <td
        className=""
        key="name"
      >
        Daniella Mante
      </td>
      <td
        className=""
        key="action_delete"
      >
        <button
          onClick={[Function]}
        >
          Delete
        </button>
      </td>
    </tr>
    <tr
      className=""
      data-key={3}
      key="3"
    >
      <td
        className=""
        key="dayOfArrival"
      >
        5 12/18
      </td>
      <td
        className=""
        key="name"
      >
        Estrella Feil
      </td>
      <td
        className=""
        key="action_delete"
      >
        <button
          onClick={[Function]}
        >
          Delete
        </button>
      </td>
    </tr>
    <tr
      className=""
      data-key={4}
      key="4"
    >
      <td
        className=""
        key="dayOfArrival"
      >
        12 22/19
      </td>
      <td
        className=""
        key="name"
      >
        Hilbert McKenzie
      </td>
      <td
        className=""
        key="action_delete"
      >
        <button
          onClick={[Function]}
        >
          Delete
        </button>
      </td>
    </tr>
    <tr
      className=""
      data-key={5}
      key="5"
    >
      <td
        className=""
        key="dayOfArrival"
      >
        9 7/15
      </td>
      <td
        className=""
        key="name"
      >
        Myah Kerluke
      </td>
      <td
        className=""
        key="action_delete"
      >
        <button
          onClick={[Function]}
        >
          Delete
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table With column descriptions should match the snapshot 1`] = `
<table>
  <thead>
    <tr>
      <th
        key="dayOfArrival"
      >
        Day Of Arrival
      </th>
      <th
        key="name"
      >
        Applicant Name
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className=""
      data-key={1}
      key="1"
    >
      <td
        className=""
        key="dayOfArrival"
      >
        1 2/18
      </td>
      <td
        className=""
        key="name"
      >
        Verla Bashirian
      </td>
    </tr>
    <tr
      className=""
      data-key={2}
      key="2"
    >
      <td
        className=""
        key="dayOfArrival"
      >
        2 16/17
      </td>
      <td
        className=""
        key="name"
      >
        Daniella Mante
      </td>
    </tr>
    <tr
      className=""
      data-key={3}
      key="3"
    >
      <td
        className=""
        key="dayOfArrival"
      >
        5 12/18
      </td>
      <td
        className=""
        key="name"
      >
        Estrella Feil
      </td>
    </tr>
    <tr
      className=""
      data-key={4}
      key="4"
    >
      <td
        className=""
        key="dayOfArrival"
      >
        12 22/19
      </td>
      <td
        className=""
        key="name"
      >
        Hilbert McKenzie
      </td>
    </tr>
    <tr
      className=""
      data-key={5}
      key="5"
    >
      <td
        className=""
        key="dayOfArrival"
      >
        9 7/15
      </td>
      <td
        className=""
        key="name"
      >
        Myah Kerluke
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table Without column descriptions should match the snapshot 1`] = `
<table>
  <thead>
    <tr>
      <th
        key="id"
      >
        Id
      </th>
      <th
        key="name"
      >
        Name
      </th>
      <th
        key="dayOfArrival"
      >
        Day Of Arrival
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className=""
      data-key={1}
      key="1"
    >
      <td
        key="id"
      >
        1
      </td>
      <td
        key="name"
      >
        Verla Bashirian
      </td>
      <td
        key="dayOfArrival"
      >
        1/2/18
      </td>
    </tr>
    <tr
      className=""
      data-key={2}
      key="2"
    >
      <td
        key="id"
      >
        2
      </td>
      <td
        key="name"
      >
        Daniella Mante
      </td>
      <td
        key="dayOfArrival"
      >
        2/16/17
      </td>
    </tr>
    <tr
      className=""
      data-key={3}
      key="3"
    >
      <td
        key="id"
      >
        3
      </td>
      <td
        key="name"
      >
        Estrella Feil
      </td>
      <td
        key="dayOfArrival"
      >
        5/12/18
      </td>
    </tr>
    <tr
      className=""
      data-key={4}
      key="4"
    >
      <td
        key="id"
      >
        4
      </td>
      <td
        key="name"
      >
        Hilbert McKenzie
      </td>
      <td
        key="dayOfArrival"
      >
        12/22/19
      </td>
    </tr>
    <tr
      className=""
      data-key={5}
      key="5"
    >
      <td
        key="id"
      >
        5
      </td>
      <td
        key="name"
      >
        Myah Kerluke
      </td>
      <td
        key="dayOfArrival"
      >
        9/7/15
      </td>
    </tr>
  </tbody>
</table>
`;
